### PR TITLE
Create block offer when non-contacts send you a message.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: src/Messages/Interactions/TSErrorMessage.m [hunk 1]
```diff
@@ -36,8 +32,14 @@ - (instancetype)initWithTimestamp:(uint64_t)timestamp
...
 
     _errorType = errorMessageType;
-    // TODO Move this out of model class.
-    [[TextSecureKitEnv sharedEnv].notificationsManager notifyUserForErrorMessage:self inThread:thread];
+    // TODO: Move this out of model class.
+    //
+    //       For now, dispatch async to ensure we're not inside a transaction
+    //       and thereby avoid deadlock.
+    TSErrorMessage *errorMessage = self;
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```
Comment (charlesmchen) [inline src/Messages/Interactions/TSErrorMessage.m:40]: `notifyUserForErrorMessage: inThread :` eventually uses a transaction within `PropertyListPreferences`, which caused deadlock if you created any `TSErrorMessage` within a transaction.

File: src/Messages/OWSUnknownContactBlockOfferMessage.m [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    return self;
+}
+
+- (nullable NSDate *)receiptDateForSorting
+{
+    // Always use date, since we're creating these interactions after the fact
+    // and back-dating them.
+    return self.date;
```
Comment (charlesmchen) [inline src/Messages/OWSUnknownContactBlockOfferMessage.m]: This isn't the most elegant solution but I feel like its better than mutating `receivedAtDate`.

**Comment by charlesmchen** [comment]
See: https://github.com/WhisperSystems/Signal-iOS/pull/1960

**Comment by charlesmchen** [comment]
Not for v2.10.

File: src/Messages/Interactions/TSErrorMessage.m [hunk 3]
```diff
@@ -47,7 +49,7 @@ - (instancetype)initWithEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
                failedMessageType:(TSErrorMessageType)errorMessageType
 {
     TSContactThread *contactThread =
-        [TSContactThread getOrCreateThreadWithContactId:envelope.source transaction:transaction];
+        [TSContactThread getOrCreateThreadWithContactId:envelope.source transaction:transaction relay:nil];
```
Comment (michaelkirk) [inline src/Messages/Interactions/TSErrorMessage.m]: why this change?

File: src/Messages/OWSUnknownContactBlockOfferMessage.m (hunk 2 above)
Comment (michaelkirk) [reply src/Messages/OWSUnknownContactBlockOfferMessage.m]: There is some nuance to our different date fields, but the comment doesn't really clear it up. Could you elaborate on the behavior this code is intended to reproduce?

**Review by michaelkirk — APPROVED** [review]
LGTM, though see my coment in Signal-iOS re: using an InfoMessage rather than an ErrorMessage.

File: src/Messages/OWSUnknownContactBlockOfferMessage.m (hunk 2 above)
Comment (charlesmchen) [reply src/Messages/OWSUnknownContactBlockOfferMessage.m]: Done.

File: src/Messages/Interactions/TSErrorMessage.m (hunk 3 above)
Comment (charlesmchen) [reply src/Messages/Interactions/TSErrorMessage.m]: Reverted.
